quickwit-storage/src/storage_with_upload_cache.rs

```diff
@@ -0,0 +1,881 @@
+    /// Lastly, the file at `{path}/[`CACHE_STATE_FILE_NAME`]` should be present with valid content.
+    pub fn from_path(
+        remote_storage: Arc<dyn Storage>,
+        storage_uri_resolver: &StorageUriResolver,
+        cache_dir: &Path,
+    ) -> StorageResult<Self> {
+        let cache_state = CacheState::from_path(cache_dir)?;
+        let cache_storage_uri = format!("file://{}", cache_dir.to_string_lossy());
+        let cache_storage = storage_uri_resolver
+            .resolve(&cache_storage_uri)
+            .map_err(|err| StorageErrorKind::InternalError.with_error(err))?;
+        let cache_storage_root = LocalFileStorage::extract_root_path_from_uri(&cache_storage_uri)?;
+
+        let num_files = cache_state.items.len();
+        let num_bytes = cache_state.items.iter().map(|(_, size)| *size).sum();
+        if cache_state
+            .disk_capacity
+            .exceeds_capacity(num_bytes, num_files)
+        {
+            return Err(StorageErrorKind::InternalError.with_error(anyhow::anyhow!(
+                "Initial cache items should not exceed cache capacity.",
+            )));
+        }
+
+        let mut disk_cache = LruCache::unbounded();
+        for (path, size) in cache_state.items {
+            disk_cache.put(path, size);
+        }
+
+        Ok(Self {
+            remote_storage,
+            cache_storage,
+            cache_storage_root,
+            disk_capacity: cache_state.disk_capacity,
+            disk_cache: Mutex::new(CacheWithMeta {
+                cache: LruCache::unbounded(),
+                num_bytes,
+                num_files,
+            }),
+        })
+    }
+
+    /// Persist the state of the entire cache.
+    ///
+    /// Takes a previously locked guard for easier mutex management.
+    async fn save_state(&self, wrapped_cache: &MutexGuard<'_, CacheWithMeta>) -> StorageResult<()> {
+        let mut cache_items = vec![];
+        for (path, size) in wrapped_cache.cache.iter() {
+            cache_items.push((path.clone(), *size));
+        }
+
+        let cache_state = CacheState {
+            disk_capacity: self.disk_capacity,
+            items: cache_items,
+        };
+
+        let file_path = self.cache_storage_root.join(CACHE_STATE_FILE_NAME);
+        let content: Vec<u8> = serde_json::to_vec(&cache_state)
+            .map_err(|err| StorageErrorKind::InternalError.with_error(err))?;
+        atomic_write(&file_path, &content)?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Storage for StorageWithUploadCache {
+    async fn put(&self, path: &Path, payload: PutPayload) -> StorageResult<()> {
+        self.remote_storage.put(path, payload.clone()).await?;
+
+        let mut locked_disk_cache = self.disk_cache.lock().await;
+        let payload_length = payload.len().await? as usize;
+        // Ignore storing an item that cannot fit in the cache.
+        if self
+            .disk_capacity
+            .exceeds_capacity(payload_length, locked_disk_cache.num_files)
```

Comment:
Did you want to write `.exceeds_capacity(payload_length, 1)`? With the current form, the logic is massively flawed, and your implementation acts as a cache for the first `max_num_files` and then does nothing:

Let assume cache capacity is 2 files, ignoring the file sizes.
file `#1`: exceeds_capacity(_, 0) -> false -> put file in cache
file `#2`: exceeds_capacity(_, 1) -> false -> put file in cache
file `#3`: exceeds_capacity(_, 2) -> true -> do nothing.

I suggest we go with something simpler:
```rust
if payload_length > self.disk_capacity.max_num_bytes {
  warn!("Failed to cache file: file size exceeds total cache capacity.");
  return Ok(())
}
```